libnavigation-core/src/test/java/com/mapbox/navigation/core/infra/TestNativeNavigator.kt

```diff
@@ -0,0 +1,184 @@
+package com.mapbox.navigation.core.infra
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.base.common.logger.Logger
+import com.mapbox.bindgen.Expected
+import com.mapbox.common.TileStore
+import com.mapbox.common.TilesetDescriptor
+import com.mapbox.navigation.base.options.DeviceProfile
+import com.mapbox.navigation.base.options.PredictiveCacheLocationOptions
+import com.mapbox.navigation.core.infra.factories.createNavigationStatus
+import com.mapbox.navigation.navigator.internal.MapboxNativeNavigator
+import com.mapbox.navigation.navigator.internal.NativeNavigatorRecreationObserver
+import com.mapbox.navigator.BannerInstruction
+import com.mapbox.navigator.CacheHandle
+import com.mapbox.navigator.ElectronicHorizonObserver
+import com.mapbox.navigator.Experimental
+import com.mapbox.navigator.FallbackVersionsObserver
+import com.mapbox.navigator.FixLocation
+import com.mapbox.navigator.GraphAccessor
+import com.mapbox.navigator.HistoryRecorderHandle
+import com.mapbox.navigator.NavigationStatus
+import com.mapbox.navigator.NavigationStatusOrigin
+import com.mapbox.navigator.NavigatorConfig
+import com.mapbox.navigator.NavigatorObserver
+import com.mapbox.navigator.PredictiveCacheController
+import com.mapbox.navigator.RoadObjectMatcher
+import com.mapbox.navigator.RoadObjectsStore
+import com.mapbox.navigator.RoadObjectsStoreObserver
+import com.mapbox.navigator.RouteAlternativesControllerInterface
+import com.mapbox.navigator.RouteInfo
+import com.mapbox.navigator.RouterError
+import com.mapbox.navigator.RouterInterface
+import com.mapbox.navigator.TilesConfig
+import io.mockk.mockk
+
+/***
+ * Emulates the behavior of the native navigator for testing.
+ *
+ * [TestNativeNavigator] doesn't cover all scenarios.
+ * If you find an empty method instead of behavior emulation feel free to implement it.
+ * Don't hesitate to write tests for this test double.
+ *
+ * Some behavior can't be easily emulated.
+ * It's hard to parse a route and emulate location to trigger the status updates.
+ * Use [TestNativeNavigator] as a stub for status update emulation.
+ * Call [updateStatus] method to trigger all observers.
+ */
+class TestNativeNavigator private constructor() : MapboxNativeNavigator {
```

Comment:
This looks huge and it will likely become only bigger if we start reusing this in another tests. 
Looks like it will even require tests on its own. 
I don't see any advantages of using this double over using mocks, which are very flexible and require much less code to write.

Reply:
Are you saying that stubs has no benefits overs mock?

Reply:
I think they are. Can you name some advantages of stubs?

Reply:
I share @Zayankovsky's sentiment as well, as we've been discussing yesterday @VysotskiVadim. Did you have a chance to try and migrate the stub factories to mock factories? How did it go?

Reply:
https://github.com/mapbox/mapbox-navigation-android/pull/5203

Reply:
merged.